File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.js
function solution(topping) {
  const cakeA = {}
  const cakeB = new Set()

  let cnt = 0;

  for (let i = 0; i < topping.length; i++) {        
      if (cakeA[topping[i]]) {
          cakeA[topping[i]]++
      } else {
          cakeA[topping[i]] = 1      
      }
  }
  let cakeATopping = Object.keys(cakeA).length;
  for (let i = 0; i < topping.length; i++) {
      cakeB.add(topping[i])
      cakeA[topping[i]]--

      if (!cakeA[topping[i]]) {
          cakeATopping--
      }
      if (cakeB.size === cakeATopping) {
          cnt+=1;
      }
  }


  return cnt;
}

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
from collections import Counter
def solution(topping):
    answer = 0
    cakeB = {}
    cakeA = Counter(topping)
    
    for i in topping:
        if i in cakeB.keys():
            cakeB[i] += 1
        else:
            cakeB[i] = 1
        cakeA[i] -= 1
        if cakeA[i] == 0:
            del cakeA[i]
        if len(cakeA.keys()) == len(cakeB.keys()):
            answer += 1
    return answer